File: packages/sdk/src/sdk/signatures/sign-transfer-locked.ts
import type { WalletClient } from 'viem'
import type { Address, Network } from '../types'
import { createDomain, TRANSFER_LOCKED_TYPES, type TransferLockedMessage } from './eip712-types'

export interface SignTransferLockedParams {
  walletClient: WalletClient
  network: Network
  verifyingContract: Address
  message: TransferLockedMessage
}

export async function signTransferLockedMessage({
  walletClient,
  network,
  verifyingContract,
  message,
}: SignTransferLockedParams): Promise<`0x${string}`> {
  const account = walletClient.account
  if (!account) {
    throw new Error('No account connected to wallet client')
  }

  const domain = createDomain(network, verifyingContract)

  const signature = await walletClient.signTypedData({
    account,
    domain,
    types: TRANSFER_LOCKED_TYPES,
    primaryType: 'TransferLocked',
    message,
  })

  return signature
}

File: sdk/py/src/flexvaults/signatures/sign_transfer_locked.py
from __future__ import annotations

from dataclasses import dataclass

from eth_account.signers.local import LocalAccount

from ..types import Address, HexString, Network
from ._signer import sign_typed_data
from .eip712_types import TRANSFER_LOCKED_TYPES, TransferLockedMessage, create_domain


@dataclass
class SignTransferLockedParams:
    account: LocalAccount
    network: Network
    verifying_contract: Address
    message: TransferLockedMessage


def sign_transfer_locked_message(params: SignTransferLockedParams) -> HexString:
    domain = create_domain(params.network, params.verifying_contract)

    message_data = {
        "userAddress": params.message.user_address,
        "toAddress": params.message.to_address,
        "lockId": params.message.lock_id,
        "amount": params.message.amount,
        "nonce": params.message.nonce,
        "serviceAddress": params.message.service_address,
    }

    return sign_typed_data(
        account=params.account,
        domain=domain,
        types=TRANSFER_LOCKED_TYPES,
        primary_type="TransferLocked",
        message=message_data,
    )
